src/main/java/org/jenkinsci/remoting/engine/Jnlp4ConnectionState.java

```
@@ -0,0 +1,83 @@
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+package org.jenkinsci.remoting.engine;
+
+import java.net.Socket;
+import java.security.cert.X509Certificate;
+import java.util.List;
+import javax.annotation.CheckForNull;
+import javax.annotation.Nonnull;
+import org.jenkinsci.remoting.protocol.impl.ConnectionRefusalException;
+
+/**
+ * Represents the connection state for a {@link JnlpProtocol4Handler} connection.
+ *
+ * @since FIXME
+ */
+public class Jnlp4ConnectionState extends JnlpConnectionState {
+
+    /**
+     * The client certificate.
+     */
+    @CheckForNull
+    private X509Certificate certificate;
+
+    /**
+     * {@inheritDoc}
+     */
+    protected Jnlp4ConnectionState(@Nonnull Socket socket,
+                                   List<? extends JnlpConnectionStateListener> listeners) {
+        super(socket, listeners);
+    }
+
+    /**
+     * Gets the client certificate (if available).
+     * @return the client certificate (if available).
+     */
+    @CheckForNull
+    public X509Certificate getCertificate() {
+        return certificate;
+    }
+
+    /**
+     * Advances the connection state to indicate that a connection has been "secured" and the property exchange
+     * is about to take place.
+     *
+     * @param certificate the client certificate.
+     * @throws ConnectionRefusalException if the connection has been refused.
+     */
+    void fireBeforeProperties(X509Certificate certificate) throws ConnectionRefusalException {
+        if (this.certificate != null) {
+            throw new IllegalStateException("fireBeforeProperties has been called already");
+        }
+        this.certificate = certificate;
+        try {
+            super.fireBeforeProperties();
+        } catch (IllegalStateException e) {
```

Nope... though there is a risk of an ISE being propagated by the super's listeners and thus causing an ISE... but that is a sign of a bad listener... and fail fast is the best way to fix those